legion/engine/physics/physx/systems/physx_physics_system.cpp

```diff
@@ -5,56 +5,96 @@ namespace legion::physics
     }
 
-    void PhysXPhysicsSystem::fixedUpdate(time::time_span<fast_time> deltaTime)
+    PhysXPhysicsSystem::~PhysXPhysicsSystem()
     {
-        m_physxScene->simulate(m_timeStep);
-        m_physxScene->fetchResults(true);
+        m_physxScene->release();
+
+        sSelfInstanceCounter--;
+
+        if (sSelfInstanceCounter == 0)
+        {
+            releasePhysXVariables();
+        }
     }
 
-    void PhysXPhysicsSystem::setupPhysXGlobalVariables()
+    void PhysXPhysicsSystem::lazyInitPhysXVariables()
     {
-        gFoundation = PxCreateFoundation(PX_PHYSICS_VERSION, gAllocator, gErrorCallback);
+        static PxDefaultAllocator sDefaultAllocator;
+        static PxDefaultErrorCallback sDefaultErrorCallback;
+
+        if (!sPhysicsStatics.m_isInit)
+        {
+            sPhysicsStatics.foundation = PxCreateFoundation(PX_PHYSICS_VERSION, sDefaultAllocator, sDefaultErrorCallback);
+
+            sPhysicsStatics.pvd = PxCreatePvd(*sPhysicsStatics.foundation);
+            PxPvdTransport* transport = PxDefaultPvdSocketTransportCreate(pvdHost, defaultPVDListeningPort, defaultPVDHostTimeout);
+            sPhysicsStatics.pvd->connect(*transport, PxPvdInstrumentationFlag::eDEBUG);
 
-        PxPvd* pvd = PxCreatePvd(*gFoundation);
-        PxPvdTransport* transport = PxDefaultPvdSocketTransportCreate(pvdHost, defaultPVDListeningPort, defaultPVDHostTimeout);
-        pvd->connect(*transport, PxPvdInstrumentationFlag::eDEBUG);
+            sPhysicsStatics.dispatcher = PxDefaultCpuDispatcherCreate(0); //deal with multithreading later on
+        }
 
-        gPhysics = PxCreatePhysics(PX_PHYSICS_VERSION, *gFoundation, PxTolerancesScale(), true, pvd);
+        if (!gPhysics)
+        {
+            gPhysics = PxCreatePhysics(PX_PHYSICS_VERSION, *sPhysicsStatics.foundation, PxTolerancesScale(), true, sPhysicsStatics.pvd);
+        }
+
+        m_defaultAllocator = &sDefaultAllocator;
+        m_defaultErrorCallback = &sDefaultErrorCallback;
     }
 
     void PhysXPhysicsSystem::setupDefaultScene()
     {
-        gDispatcher = PxDefaultCpuDispatcherCreate(0); //deal with multithreading later on
-
         PxSceneDesc sceneDesc(gPhysics->getTolerancesScale());
         sceneDesc.gravity = PxVec3(0.0f, -9.81f, 0.0f);
-        sceneDesc.cpuDispatcher = gDispatcher;
+        sceneDesc.cpuDispatcher = sPhysicsStatics.dispatcher;
         sceneDesc.filterShader = PxDefaultSimulationFilterShader;
         sceneDesc.flags |= PxSceneFlag::eENABLE_STABILIZATION;
 
         m_physxScene = gPhysics->createScene(sceneDesc);
     }
+
+    void PhysXPhysicsSystem::releasePhysXVariables()
+    {
+        sPhysicsStatics.dispatcher->release();
+        gPhysics->release();
+
+        if (sPhysicsStatics.pvd)
+        {
+            PxPvdTransport* transport = sPhysicsStatics.pvd->getTransport();
+            sPhysicsStatics.pvd->release();
+            sPhysicsStatics.pvd = nullptr;
```

Comment:
is there a specific reason for things being called `s...` or `g...` instead of the normal `m_...`?

Reply:
Mostly because I assumed m_ is only for private member variables of classes, I thought we could use something else for statics and globals. I can easily change this of course.